File: server/wwwroot/js/Awareness.js

import { Union } from "./fable_modules/fable-library-js.5.13.0/Types.js";
import { union_type, option_type, string_type, float64_type } from "./fable_modules/fable-library-js.5.13.0/Reflection.js";
import { Vec2, Vec2_$reflection } from "./Types.js";
import { defaultOf } from "./fable_modules/fable-library-js.5.13.0/Util.js";
import { Operators_IsNull } from "./fable_modules/fable-library-js.5.13.0/FSharp.Core.js";

export class Incoming extends Union {
    constructor(tag, fields) {
        super();
        this.tag = tag;
        this.fields = fields;
    }
    cases() {
        return ["Presence", "Bye"];
    }
}

export function Incoming_$reflection() {
    return union_type("Client.Awareness.Incoming", [], Incoming, () => [[["Item1", float64_type], ["name", string_type], ["color", string_type], ["cursor", option_type(Vec2_$reflection())]], [["Item", float64_type]]]);
}

export function encodePresence(id, name, color, cursor) {
    let c;
    const value = {
        type: "presence",
        id: id,
        name: name,
        color: color,
        cursor: (cursor == null) ? defaultOf() : ((c = cursor, {
            x: c.X,
            y: c.Y,
        })),
    };
    return JSON.stringify(value);
}

export function tryDecode(json) {
    try {
        const o = JSON.parse(json);
        const kind = o.type;
        switch (kind) {
            case "presence": {
                const id = o.id;
                const name = o.name;
                const color = o.color;
                const cursorRaw = o.cursor;
                return new Incoming(/* Presence */ 0, [id, name, color, Operators_IsNull(cursorRaw) ? undefined : (new Vec2(cursorRaw.x, cursorRaw.y))]);
            }
            case "bye":
                return new Incoming(/* Bye */ 1, [o.id]);
            default:
                return undefined;
        }
    }
    catch (matchValue) {
        return undefined;
    }
}

File: client/Awareness.fs
/// Encoding/decoding for the small JSON "who's online, where's your cursor" protocol
/// that rides alongside the binary Yjs updates on the same WebSocket (see
/// server/Rooms.fs and Doc.fs). Kept as plain, readable JSON on purpose - unlike the
/// Yjs traffic, this is meant to be easy to eyeball in the debug panel/console.
module Client.Awareness

open Fable.Core
open Fable.Core.JsInterop
open Client.Types

type Incoming =
    | Presence of ClientId * name: string * color: string * cursor: Vec2 option * editing: EditingRef option
    | Bye of ClientId

let encodePresence
    (id: ClientId)
    (name: string)
    (color: string)
    (cursor: Vec2 option)
    (editing: EditingRef option)
    : string =
    let cursorObj: obj =
        match cursor with
        | Some c -> createObj [ "x" ==> c.X; "y" ==> c.Y ]
        | None -> null

    let editingObj: obj =
        match editing with
        | Some(noteId, field) -> createObj [ "noteId" ==> noteId; "field" ==> field ]
        | None -> null

    createObj
        [ "type" ==> "presence"
          "id" ==> id
          "name" ==> name
          "color" ==> color
          "cursor" ==> cursorObj
          "editing" ==> editingObj ]
    |> JS.JSON.stringify

let tryDecode (json: string) : Incoming option =
    try
        let o = JS.JSON.parse json
        let kind: string = o?``type``

        match kind with
        | "presence" ->
            let id: ClientId = o?id
            let name: string = o?name
            let color: string = o?color
            let cursorRaw: obj = o?cursor

            let cursor =
                if isNull cursorRaw then
                    None
                else
                    Some { X = cursorRaw?x; Y = cursorRaw?y }

            let editingRaw: obj = o?editing

            let editing =
                if isNull editingRaw then
                    None
                else
                    Some(editingRaw?noteId, editingRaw?field)

            Some(Presence(id, name, color, cursor, editing))
        | "bye" ->
            let id: ClientId = o?id
            Some(Bye id)
        | _ -> None
    with _ ->
        None
